crates/ntx-builder/src/actor.rs

```diff
@@ -0,0 +1,176 @@
+    /// Address of the remote prover. If `None`, transactions will be proven locally, which is
+    // undesirable due to the performance impact.
+    pub tx_prover_url: Option<Url>,
+    /// The latest chain state that account actors can rely on.
+    pub chain_state: Arc<RwLock<ChainState>>,
+}
+
+#[derive(Default, Debug)]
+enum ActorMode {
+    #[default]
+    AwaitingEvents,
+    ExecutingTransactions,
+    AwaitingCommit(TransactionId),
+}
+
+/// Account actor that manages state and processes transactions for a single network account.
+pub struct AccountActor {
+    state: ActorMode,
+    event_rx: mpsc::UnboundedReceiver<MempoolEvent>,
+    block_producer: BlockProducerClient,
+    prover: Option<RemoteTransactionProver>,
+    chain_state: Arc<RwLock<ChainState>>,
+}
+
+impl AccountActor {
+    /// Constructs a new account actor and corresponding messaging channel with the given
+    /// configuration.
+    pub fn new(config: &AccountActorConfig) -> (Self, mpsc::UnboundedSender<MempoolEvent>) {
+        let block_producer = BlockProducerClient::new(config.block_producer_url.clone());
+        let prover = config.tx_prover_url.clone().map(RemoteTransactionProver::new);
+        let (event_tx, event_rx) = mpsc::unbounded_channel();
+        let actor = Self {
+            state: ActorMode::default(),
+            event_rx,
+            block_producer,
+            prover,
+            chain_state: config.chain_state.clone(),
+        };
+        (actor, event_tx)
+    }
+
+    /// Runs the account actor, processing events and managing state until a reason to shutdown is
+    /// encountered.
+    pub async fn run(mut self, mut state: State, semaphore: Arc<Semaphore>) -> ActorShutdownReason {
+        loop {
+            // Enable or disable transaction execution based on actor mode.
+            let tx_execution_fut = match self.state {
+                // Disable transaction execution.
+                ActorMode::AwaitingEvents | ActorMode::AwaitingCommit(_) => {
+                    std::future::pending().boxed()
+                },
+                // Enable transaction execution.
+                ActorMode::ExecutingTransactions => semaphore.acquire().boxed(),
+            };
+            tokio::select! {
+                // Handle mempool events.
+                event = self.event_rx.recv() => {
+                    let Some(event) = event else {
+                         return ActorShutdownReason::EventChannelClosed;
+                    };
+                    // Re-enable transaction execution if the transaction being waited on has been
+                    // committed.
+                    if let ActorMode::AwaitingCommit(tx_id) = self.state {
+                        if let MempoolEvent::BlockCommitted { txs, .. } = &event {
+                            if txs.contains(&tx_id) {
+                                self.state = ActorMode::ExecutingTransactions;
+                            }
+                        }
```

Comment:
This waits until the tx has been committed, but it should be enough to wait for it to be accepted by the mempool i.e. `MempoolEvent::TransactionAdded`